File: python/resdata/rd_type.py
from cwrap import BaseCClass, BaseCEnum

import resdata.types._type as _rd_type


class ResdataTypeEnum(BaseCEnum):
    TYPE_NAME = "rd_type_enum"
    RD_CHAR_TYPE = None
    RD_FLOAT_TYPE = None
    RD_DOUBLE_TYPE = None
    RD_INT_TYPE = None
    RD_BOOL_TYPE = None
    RD_MESS_TYPE = None
    RD_STRING_TYPE = None


ResdataTypeEnum.addEnum("RD_CHAR_TYPE", 0)
ResdataTypeEnum.addEnum("RD_FLOAT_TYPE", 1)
ResdataTypeEnum.addEnum("RD_DOUBLE_TYPE", 2)
ResdataTypeEnum.addEnum("RD_INT_TYPE", 3)
ResdataTypeEnum.addEnum("RD_BOOL_TYPE", 4)
ResdataTypeEnum.addEnum("RD_MESS_TYPE", 5)
ResdataTypeEnum.addEnum("RD_STRING_TYPE", 7)


class ResDataType(BaseCClass):
    TYPE_NAME = "rd_data_type"

    def __init__(self, type_enum=None, element_size=None, type_name=None):
        self._assert_valid_arguments(type_enum, element_size, type_name)

        if type_name:
            c_ptr = _rd_type._create_from_name(type_name)
        elif element_size is None:
            c_ptr = _rd_type._create_from_type(type_enum)
        else:
            c_ptr = _rd_type._create(type_enum, element_size)

        super().__init__(c_ptr)

    def _assert_valid_arguments(self, type_enum, element_size, type_name):
        if type_name is not None:
            if type_enum is not None or element_size is not None:
                err_msg = (
                    "Type name given (%s). Expected both "
                    + "type_enum and element_size to be None"
                )
                raise ValueError(err_msg % type_name)

        elif type_enum is None:
            raise ValueError("Both type_enum and type_name is None!")

        elif type_enum == ResdataTypeEnum.RD_STRING_TYPE:
            if element_size is None:
                raise ValueError(
                    "When creating an RD_STRING one must " + "provide an element size!"
                )

            if not (0 <= element_size <= 999):
                raise ValueError(
                    "Expected element_size to be in the range "
                    + "[0, 999], was: %d" % element_size
                )

    @property
    def type(self):
        return ResdataTypeEnum(_rd_type._get_type(self))

    @property
    def element_size(self):
        return _rd_type._get_sizeof_iotype(self)

    @property
    def type_name(self):
        return _rd_type._name(self)

    def free(self):
        _rd_type._free(self)

    def is_int(self):
        return _rd_type._is_int(self)

    def is_char(self):
        return _rd_type._is_char(self)

    def is_float(self):
        return _rd_type._is_float(self)

    def is_double(self):
        return _rd_type._is_double(self)

    def is_mess(self):
        return _rd_type._is_mess(self)

    def is_bool(self):
        return _rd_type._is_bool(self)

    def is_string(self):
        return _rd_type._is_string(self)

    def is_numeric(self):
        return _rd_type._is_numeric(self)

    def is_equal(self, other):
        return _rd_type._is_equal(self, other)

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.is_equal(other)
        return False

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return hash((self.type, self.element_size))

    @classmethod
    def create_from_type_name(cls, name):
        return ResDataType(type_name=name)

    # Enables one to fetch a type as ResDataType.RD_XXXX
    class classproperty:
        def __init__(self, fget):
            self.fget = fget

        def __get__(self, owner_self, owner_cls):
            return self.fget(owner_cls)

    @classproperty
    def RD_INT(cls):
        return ResDataType(ResdataTypeEnum.RD_INT_TYPE)

    @classproperty
    def RD_FLOAT(cls):
        return ResDataType(ResdataTypeEnum.RD_FLOAT_TYPE)

    @classproperty
    def RD_DOUBLE(cls):
        return ResDataType(ResdataTypeEnum.RD_DOUBLE_TYPE)

    @classproperty
    def RD_BOOL(cls):
        return ResDataType(ResdataTypeEnum.RD_BOOL_TYPE)

    @classproperty
    def RD_MESS(cls):
        return ResDataType(ResdataTypeEnum.RD_MESS_TYPE)

    @classproperty
    def RD_CHAR(cls):
        return ResDataType(ResdataTypeEnum.RD_CHAR_TYPE)

    @classmethod
    def RD_STRING(cls, elem_size):
        return ResDataType(ResdataTypeEnum.RD_STRING_TYPE, elem_size)

File: lib/resdata/rd_type.cpp
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <ert/util/util.hpp>
#include <resdata/rd_type.hpp>

#define RD_TYPE_NAME_CHAR "CHAR"
#define RD_TYPE_NAME_FLOAT "REAL"
#define RD_TYPE_NAME_INT "INTE"
#define RD_TYPE_NAME_DOUBLE "DOUB"
#define RD_TYPE_NAME_BOOL "LOGI"
#define RD_TYPE_NAME_MESSAGE "MESS"

static char *alloc_string_name(const rd_data_type rd_type) {
    return util_alloc_sprintf("C%03d", rd_type_get_sizeof_iotype(rd_type));
}

static bool is_rd_string_name(const char *type_name) {
    return (type_name[0] == 'C' && isdigit(type_name[1]) &&
            isdigit(type_name[2]) && isdigit(type_name[3]));
}

static size_t get_rd_string_length(const char *type_name) {
    if (!is_rd_string_name(type_name))
        util_abort("%s: Expected eclipse string (CXXX), received %s\n",
                   __func__, type_name);

    return atoi(type_name + 1);
}

rd_data_type rd_type_create(const rd_type_enum type,
                            const size_t element_size) {
    rd_data_type rd_type =
        (type == RD_STRING_TYPE ? RD_STRING(element_size)
                                : rd_type_create_from_type(type));

    if ((size_t)rd_type_get_sizeof_iotype(rd_type) != element_size)
        util_abort(
            "%s: element_size mismatch for type %d, was: %d, expected: %d\n",
            __func__, type, element_size, rd_type_get_sizeof_iotype(rd_type));

    return rd_type;
}

rd_data_type rd_type_create_from_type(const rd_type_enum type) {
    switch (type) {
    case (RD_CHAR_TYPE):
        return RD_CHAR;
    case (RD_INT_TYPE):
        return RD_INT;
    case (RD_FLOAT_TYPE):
        return RD_FLOAT;
    case (RD_DOUBLE_TYPE):
        return RD_DOUBLE;
    case (RD_BOOL_TYPE):
        return RD_BOOL;
    case (RD_MESS_TYPE):
        return RD_MESS;
    case (RD_STRING_TYPE):
        util_abort("%s: Variable length string type cannot be created"
                   " from type alone!\n",
                   __func__);
        return RD_STRING(0); /* Dummy */
    default:
        util_abort("%s: invalid rd_type: %d\n", __func__, type);
        return RD_INT; /* Dummy */
    }
}

rd_type_enum rd_type_get_type(const rd_data_type rd_type) {
    return rd_type.type;
}

char *rd_type_alloc_name(const rd_data_type rd_type) {
    switch (rd_type.type) {
    case (RD_CHAR_TYPE):
        return util_alloc_string_copy(RD_TYPE_NAME_CHAR);
    case (RD_STRING_TYPE):
        return alloc_string_name(rd_type);
    case (RD_FLOAT_TYPE):
        return util_alloc_string_copy(RD_TYPE_NAME_FLOAT);
    case (RD_DOUBLE_TYPE):
        return util_alloc_string_copy(RD_TYPE_NAME_DOUBLE);
    case (RD_INT_TYPE):
        return util_alloc_string_copy(RD_TYPE_NAME_INT);
    case (RD_BOOL_TYPE):
        return util_alloc_string_copy(RD_TYPE_NAME_BOOL);
    case (RD_MESS_TYPE):
        return util_alloc_string_copy(RD_TYPE_NAME_MESSAGE);
    default:
        util_abort("Internal error in %s - internal eclipse_type: %d not "
                   "recognized - aborting \n",
                   __func__, rd_type.type);
        return NULL; /* Dummy */
    }
}

rd_data_type rd_type_create_from_name(const char *type_name) {
    if (strncmp(type_name, RD_TYPE_NAME_FLOAT, RD_TYPE_LENGTH) == 0)
        return RD_FLOAT;
    else if (strncmp(type_name, RD_TYPE_NAME_INT, RD_TYPE_LENGTH) == 0)
        return RD_INT;
    else if (strncmp(type_name, RD_TYPE_NAME_DOUBLE, RD_TYPE_LENGTH) == 0)
        return RD_DOUBLE;
    else if (strncmp(type_name, RD_TYPE_NAME_CHAR, RD_TYPE_LENGTH) == 0)
        return RD_CHAR;
    else if (is_rd_string_name(type_name))
        return RD_STRING(get_rd_string_length(type_name));
    else if (strncmp(type_name, RD_TYPE_NAME_MESSAGE, RD_TYPE_LENGTH) == 0)
        return RD_MESS;
    else if (strncmp(type_name, RD_TYPE_NAME_BOOL, RD_TYPE_LENGTH) == 0)
        return RD_BOOL;
    else {
        util_abort("%s: unrecognized type name:%s \n", __func__, type_name);
        return RD_INT; /* Dummy */
    }
}

int rd_type_get_sizeof_ctype(const rd_data_type rd_type) {
    return rd_type.element_size;
}

int rd_type_get_sizeof_iotype(const rd_data_type rd_type) {
    if (rd_type_is_bool(rd_type))
        return sizeof(int);

    if (rd_type_is_char(rd_type))
        return rd_type.element_size - 1;

    if (rd_type_is_string(rd_type))
        return rd_type.element_size - 1;

    return rd_type.element_size;
}

bool rd_type_is_numeric(const rd_data_type rd_type) {
    return (rd_type_is_int(rd_type) || rd_type_is_float(rd_type) ||
            rd_type_is_double(rd_type));
}

bool rd_type_is_alpha(const rd_data_type rd_type) {
    return (rd_type_is_char(rd_type) || rd_type_is_mess(rd_type) ||
            rd_type_is_string(rd_type));
}

bool rd_type_is_equal(const rd_data_type rd_type1,
                      const rd_data_type rd_type2) {
    return (rd_type1.type == rd_type2.type &&
            rd_type1.element_size == rd_type2.element_size);
}

bool rd_type_is_char(const rd_data_type rd_type) {
    return (rd_type.type == RD_CHAR_TYPE);
}

bool rd_type_is_int(const rd_data_type rd_type) {
    return (rd_type.type == RD_INT_TYPE);
}

bool rd_type_is_float(const rd_data_type rd_type) {
    return (rd_type.type == RD_FLOAT_TYPE);
}

bool rd_type_is_double(const rd_data_type rd_type) {
    return (rd_type.type == RD_DOUBLE_TYPE);
}

bool rd_type_is_mess(const rd_data_type rd_type) {
    return (rd_type.type == RD_MESS_TYPE);
}

bool rd_type_is_bool(const rd_data_type rd_type) {
    return (rd_type.type == RD_BOOL_TYPE);
}

bool rd_type_is_string(const rd_data_type rd_type) {
    return (rd_type.type == RD_STRING_TYPE);
}
